File: Recursion/MergeSort.java
package Recursion;

import java.util.Arrays;

/**
 * MergeSort
 */
public class MergeSort {

    public static void main(String[] args) {
        // randome array of integers
        int[] arr = { 5, 4, 3, 2, 1 };
        // print unsorted array
        System.out.println("Unsorted array: ");
        System.out.println(Arrays.toString(arr));
        // sort array
        mergeSort(arr, 0, arr.length - 1);
        // print sorted array
        System.out.println("Sorted array: ");
        System.out.println(Arrays.toString(arr));
    }

    static void mergeSort(int[] arr, int start, int end) {
        if (start < end) {
            int mid = ((start + end) / 2);
            mergeSort(arr, start, mid);
            mergeSort(arr, mid + 1, end);
            merge(arr, start, mid, end);
        }
    }
    // merge two 
    static void merge(int[] arr, int start, int mid, int end) {
        int[] temp = new int[end - start + 1];
        int i = start;
        int j = mid + 1;
        int k = 0;
        while (i <= mid && j <= end) {
            if (arr[i] < arr[j]) {
                temp[k] = arr[i++];
            } else {
                temp[k] = arr[j++];
            }
            k++;
        }
        // this is for thie case of the array is not sorted;
        while (i <= mid) {
            temp[k++] = arr[i++];
        }   
        while (j <= end) {
            temp[k++] = arr[j++];
        }
        for (int l = 0; l < temp.length; l++) {
            arr[start + l] = temp[l];
        }
    }
}

File: Recursion/sorts/mergesort.js
// make a recursive mergesort alog
var bubbleSort = (arr) => {
    // console.log(arr);
    for (var j = 0; j < arr.length; j++) {
        for (var i = 0; i < arr.length; i++) {
            if (arr[i] > arr[i + 1]) {
                var temp = arr[i + 1];
                arr[i + 1] = arr[i];
                arr[i] = temp;
                // console.log("true")
            }
            // console.log("false")
        }
        // console.log(arr)
        console.log(j);
    }
}
var arr = [4, 3, 7, 6, 4, 2, 8];
// mergesort
var mergeSort = (arr, start, end ) => {
    if (start < end) {
        var mid = Math.floor((start + end) / 2);
        mergeSort(arr, start, mid);
        mergeSort(arr, mid + 1, end);
        merge(arr, start, mid, end);
    }
}
var merge = (arr, start, mid, end) => {
    var temp = [];
    var i = start;
    var j = mid + 1;
    var k = 0;
    while (i <= mid && j <= end) {
        if (arr[i] < arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }
    while (i <= mid) {
        temp[k++] = arr[i++];
    }
    while (j <= end) {
        temp[k++] = arr[j++];
    }
    for (var i = start; i <= end; i++) {
        arr[i] = temp[i - start];
    }
}
mergeSort(arr, 0, arr.length - 1);
console.log(arr);